File: js/src/core/timeline-view.ts
import { BaseView } from './base-view';
import { TimelineContent, TimelineItem, TimelineStatus } from '../types';
import { MissingRequiredParameterError } from '../errors';

// TimelineView captures chronological progress such as onboarding steps or activity feeds.

/**
 * Builds a Timeline SGUI view — a chronological list of events such as
 * onboarding steps or activity feeds.
 *
 * Entries are appended via `addItem`/`addEvent` or replaced wholesale with
 * `setItems`; `setIntro` adds lead context text.
 *
 * Extends {@link BaseView}; instantiated by the YeriaApp/YeriaUI factory,
 * populated with these builders, then serialized to a JSON view description and
 * signed into a v3 envelope by `serve()`.
 */
export class TimelineView extends BaseView {

    static fromJson(json: Record<string, unknown>): TimelineView {
        return TimelineView.fromJsonAs(TimelineView, 'Timeline', json);
    }
    constructor(viewId: string, title: string, processId?: string) {
        super({
            id: viewId,
            type: 'Timeline',
            processId,
            metadata: {
                version: '1.0.0',
                createdAt: new Date()
            }
        });

        this.content = {
            title,
            intro: '',
            items: []
        } as TimelineContent;
    }

    // Adds optional context text displayed before the timeline items.
    setIntro(intro: string): this {
        return this.setIntroText('intro', intro);
    }

    // Persists a fully configured timeline entry.
    addItem(item: TimelineItem): this {
        if (!item.id || !item.title || !item.timestamp) {
            throw new MissingRequiredParameterError('timeline item id, title, and timestamp');
        }

        (this.content as TimelineContent).items.push({
            ...item,
            id: item.id.trim(),
            title: item.title.trim(),
            timestamp: item.timestamp.trim(),
            description: item.description?.trim()
        });

        return this;
    }

    // Convenience builder that assembles a timeline entry from primitive fields and appends it.
    addEvent(
        id: string,
        title: string,
        timestamp: string,
        options: { description?: string; status?: TimelineStatus; icon?: string } = {}
    ): this {
        return this.addItem({
            id,
            title,
            timestamp,
            description: options.description,
            status: options.status,
            icon: options.icon
        });
    }

    // Replaces all events at once, enforcing validation per entry.
    setItems(items: TimelineItem[]): this {
        (this.content as TimelineContent).items = [];
        items.forEach(item => this.addItem(item));
        return this;
    }

    clearItems(): this {
        (this.content as TimelineContent).items = [];
        return this;
    }

    // Gives access to the final serializable data.
    getContent(): TimelineContent {
        return this.content as TimelineContent;
    }
}

File: py/yeriasdk/views/timeline_view.py
"""
TimelineView - A view for displaying chronological progress
"""

from typing import Optional
from datetime import datetime

from ..core.base_view import BaseView
from ..types.models import TimelineItem, TimelineStatus
from ..errors.exceptions import MissingRequiredParameterError


class TimelineView(BaseView):
    """View for displaying timeline items"""

    def __init__(self, view_id: str, title: str, process_id: Optional[str] = None):
        super().__init__(
            {
                "id": view_id,
                "type": "Timeline",
                "process_id": process_id,
                "metadata": {
                    "version": "1.0.0",
                    "created_at": datetime.now(),
                },
            }
        )

        self.content = {
            "title": title,
            "intro": "",
            "items": [],
        }

    def set_intro(self, intro: str) -> "TimelineView":
        """Add optional context text displayed before the timeline items"""
        return self._set_intro_text("intro", intro)

    def add_item(self, item: TimelineItem) -> "TimelineView":
        """Persist a fully configured timeline entry"""
        if not item.id or not item.title or not item.timestamp:
            raise MissingRequiredParameterError(
                "timeline item id, title, and timestamp"
            )

        item_dict = {
            "id": item.id.strip(),
            "title": item.title.strip(),
            "timestamp": item.timestamp.strip(),
            "description": item.description.strip() if item.description else None,
            "status": item.status,
            "icon": item.icon,
        }

        if item.meta:
            item_dict["meta"] = item.meta

        self.content["items"].append(item_dict)
        return self

    def add_event(
        self,
        id: str,
        title: str,
        timestamp: str,
        description: Optional[str] = None,
        status: Optional[TimelineStatus] = None,
        icon: Optional[str] = None,
    ) -> "TimelineView":
        """Add an event to the timeline"""
        item = TimelineItem(
            id=id,
            title=title,
            timestamp=timestamp,
            description=description,
            status=status,
            icon=icon,
        )
        return self.add_item(item)

    def set_items(self, items: list) -> "TimelineView":
        """Replace all events at once, enforcing validation per entry"""
        self.content["items"] = []
        for item in items:
            if isinstance(item, dict):
                timeline_item = TimelineItem(
                    id=item["id"],
                    title=item["title"],
                    timestamp=item["timestamp"],
                    description=item.get("description"),
                    status=item.get("status"),
                    icon=item.get("icon"),
                    meta=item.get("meta"),
                )
                self.add_item(timeline_item)
            else:
                self.add_item(item)
        return self

    def clear_items(self) -> "TimelineView":
        """Clear all items"""
        self.content["items"] = []
        return self

    def get_content(self):
        """Get the timeline content"""
        return self.content

